### PR TITLE
Nf mtms csd model

### PR DESCRIPTION
**Pull request by MrBago**
I've implemented the multi-tissue multi-shell CSD model and wanted to push it upstream.

### CODE REVIEW

**Comment by MrBago** [comment]
I realized that I left out the code for making the response function. For my project I was building the response function from using freesurfer segmentation. I've put that code into a gist, https://gist.github.com/MrBago/a2e141a0b2a9bcb772e4e59a22daefba.

Also when I use this model, I also use the parallel framework (https://github.com/nipy/dipy/pull/642) to make it faster. I believe @sahmed95 is working on getting that work ready to merge.

**Comment by MrBago** [comment]
Hi all. I know there is a release coming up and I think this is a really solid model we should include in the next release if we can. Let me know if there is anything I can do to help streamline the review of this PR.

**Comment by arokem** [comment]
Hey @MrBago : one way to help is to take a look at https://github.com/nipy/dipy/pull/1180, which is exploring a possible replacement for cvxopt. This will also affect what you are doing here, obviously.

**Comment by arokem** [comment]
Hey @MrBago : any more thoughts about my comment (using cvxpy instead of using cvxopt)?

**Comment by MrBago** [comment]
I should have time next weekend and I think I can get it done then.

**Comment by skoudoro** [comment]
[#1168](https://trello.com/c/ZgmjR1IN/18-1168)

**Comment by GuillaumeTh** [comment]
Hi all, I have a problem with the PR, I test it with the peaks_from_model function and the computing is really long (more than 24 hours for a dwi in 1mm isotropic).

I have test in 2 processes and it's more fast than 8 processes. I think we have a problem with the multithreading. Maybe a variable in concurency.

Moreover, we must to set the variable OMP_NUM_THREADS to 1 or another number else cvxopt create other threads like numpy.

**Comment by MrBago** [comment]
Yes this model can be very resource hungry, especially if you're using 1mm
isotropic X many gradient directions. I used this model on some large data
with https://github.com/nipy/dipy/pull/642 and got the compute time to < 6
hours with 4 cores I think. I'd suggest setting MKL/OMP num threads to 1.

Bago

**Comment by MrBago** [comment]
For context, my experience was using "human connectome project" data, 1.25
mm3, 270 gradient directions. I'd also recommend spending some time making
sure you've got a good linear algebra library linked up. I was using the
MKL version packaged with anaconda, which I found to be the fastest on the
machines I was using at the time.

**Comment by Garyfallidis** [comment]
The obvious question here is how much the original implementation takes and what libraries are used there (if faster). The second question is where is the actual bottleneck (what takes most of the time).
Third question is where is the tutorial to help the reviewing process. 

It would be great to have this merged for the next release. So, let's try to address the different issues here.

**Comment by GuillaumeTh** [comment]
Hi guys, I try to compute msmt fodf on a data which have a 0.3 mm isotropic of voxel size and I have an error. Have you any ideas ?

Thanks

**Comment by MrBago** [comment]
Either my code has a bug, or maybe you did something wrong :).

If you post the trace & an MCVE (https://stackoverflow.com/help/mcve) I
might be able to be more helpful.

**Comment by arokem** [comment]
BTW - any chance to rewrite this using cvxpy?

**Comment by GuillaumeTh** [comment]
@MrBago Sorry ! It was an obvious error with my Response Function. 

@arokem I can help to test the code. But I can't refactor the code with cvxpy.

**Comment by GuillaumeTh** [comment]
@MrBago Is it ok if me and @jchoude create some PR on your branch to merge this PR ?

@arokem I will check to rewrite using cvxpy

**Comment by MrBago** [comment]
@GuillaumeTh sorry I didn't get around to this sooner, feel free to make a PR against this branch or make a fresh PR against dipy. Let me know if I can be of any help, I'll try and make time to review if it'll be useful.

**Comment by GuillaumeTh** [comment]
@arokem Have you an idea about the solver that I need to use for the fitting ?

**Comment by GuillaumeTh** [comment]
Hi guys,

I refactor the code with cvxpy but I have a big problem:

If I use ECOS as solver, the unit test work but there is a problem with large scale number (https://github.com/cvxgrp/cvxpy/issues/261, https://github.com/embotech/ecos/issues/142). So, I can't solve for fodf.

If I use SCS as solver, the unit test didn't work because the difference with the ground truth is 2 decimals. 

Want do you think about that ? I know it's important to use cvxpy but now I'm stuck

**Comment by MrBago** [comment]
@arokem I saw that on #1180 there was some conversation about the licensing of the two solvers. Do you know of the top of your head are both the ECOS & SCS solvers compatible with the dipy licence?

**Comment by arokem** [comment]
I think either are fine.

**Comment by GuillaumeTh** [comment]
Hi,
Have you an idea how to refactor the fitting part ?

@rutgerfick have you an idea ? I saw your PR on mapmri...

**Comment by rutgerfick** [comment]
Hi @GuillaumeTh, I am not familiar with this code. What is exactly the problem and in which lines?

**Comment by GuillaumeTh** [comment]
In msd.py at line 218, the ECOS solver doesn't solve the problem when the number are to high (higher than 10e9).

Is it normal ?

**Comment by rutgerfick** [comment]
Hi @GuillaumeTh, I'm not 100% sure what is happening exactly but my I'm guessing the number you are talking about is the cost function value?

If this is the case then I think your optimization is doing something you're not intending. Is it possible that your spherical harmonics coefficients are exploding due to lack of sufficient regularization?

**Comment by GuillaumeTh** [comment]
Hi @rutgerfick I think cvxpy is not robust to do this type of solving. I tried other solver with the same data and they worked. Yesterday I tested OSQP a solver with an Apache license (like cvxpy) and the results are good and the computing time is better than cvxopt. 

So I can refactor the code with OSQP. @arokem @Garyfallidis is it possible to use that ?

**Comment by rutgerfick** [comment]
Hi @GuillaumeTh , perhaps a bit of a late response but I implemented the MSMT-CSD algorithm in Dmipy using the cvxpy solver and it seems to do a good job. Take a look at the [code](https://github.com/AthenaEPI/dMipy) and  [example](http://nbviewer.jupyter.org/github/AthenaEPI/dmipy/blob/master/examples/example_multi_tissue_constrained_spherical_deconvolution.ipynb) if you're interested.

**Comment by arokem** [comment]
@ShreyasFadnavis : are you following up on this work? Could you open a new PR from your branch of this and then close this?

File: dipy/reconst/msd.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+from dipy.reconst import csdeconv as csd
+from dipy.reconst.multi_voxel import multi_voxel_fit
+
+from ..utils.optpkg import optional_package
+
+cvx, have_cvxopt, _ = optional_package("cvxopt")
+if have_cvxopt:
+    cvx.solvers.options['show_progress'] = False
+
+sh_const = .5 / np.sqrt(np.pi)
```
Comment (Borda) [inline dipy/reconst/msd.py]: looks like a constant, then rather use `SH_CONST = .5 / np.sqrt(np.pi)`

File: dipy/reconst/msd.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    r = cvx.solvers.lp(c, G, h)
+    x = np.asarray(r['x'])[:, 0]
+    out = np.zeros(B.shape[1])
+    out[n == 0] = sh_const
+    out[n != 0] = x
+
+    iso_d = [sh_const] * iso
+    return np.concatenate([iso_d, out])
+
+delta_functions = {"basic":_basic_delta,
```
Comment (Borda) [inline dipy/reconst/msd.py]: also probably a constant like...

**Comment by ShreyasFadnavis** [comment]
> @ShreyasFadnavis : are you following up on this work? Could you open a new PR from your branch of this and then close this?

Hi @arokem,

Yes, I will be following up on this work! I am facing some issues with porting a few lines to CVXPY but will have it done in the coming days.

You can go ahead and close this PR! I will create a new one soon!

**Comment by ShreyasFadnavis** [comment]
> I am missing any documentation or comments, the definition of inputs/outputs

Hello @Borda,

Thank you for these suggestions. I will incorporate these in my PR!

**Comment by ShreyasFadnavis** [comment]
Thanks @arokem! Opening up a new PR shortly :+1:

**Pull request closed without merging**